File: src/main/java/xyz/oribuin/eternalreports/utils/OriCommand.java
package xyz.oribuin.eternalreports.utils;

import org.bukkit.Bukkit;
import org.bukkit.command.*;

import java.util.List;

public abstract class OriCommand implements TabExecutor {

    private final String commandName;

    public OriCommand(String commandName) {
        this.commandName = commandName;
    }

    // In Theory: Register the command.
    public void registerCommand() {
        PluginCommand cmd = Bukkit.getPluginCommand(this.getName());

        if (cmd != null) {
            cmd.setExecutor(this);
            cmd.setTabCompleter(this);
        }
    }

    // Register Command
    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        return false;
    }


    // Register command tab complete
    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String alias, String[] args) {
        return null;
    }

    // Get Command Name
    public String getName() {
        return commandName;
    }
}

File: src/main/kotlin/xyz/oribuin/eternalreports/commands/OriCommand.kt
package xyz.oribuin.eternalreports.commands

import org.bukkit.Bukkit
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.command.TabExecutor
import xyz.oribuin.eternalreports.EternalReports

abstract class OriCommand(open val plugin: EternalReports, // Get Command Name
                          private val name: String) : TabExecutor {
    fun registerCommand() {
        val cmd = Bukkit.getPluginCommand(name)
        if (cmd != null) {
            cmd.setExecutor(this)
            cmd.tabCompleter = this
        }
    }

    abstract fun executeCommand(sender: CommandSender, args: Array<String>)

    abstract fun tabComplete(sender: CommandSender, args: Array<String>): List<String>?

    // Register Command
    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<String>): Boolean {
        executeCommand(sender, args)
        return true
    }

    // Register command tab complete
    override fun onTabComplete(sender: CommandSender, command: Command, alias: String, args: Array<String>): List<String>? {
        return tabComplete(sender, args)
    }

    val command: Command?
        get() = plugin.getCommand(name)

}